### ISSUE TITLE
General Query

### ISSUE THREAD

**Issue by LingarajVB** [nicohaenggi/SafariBooks-Downloader#47]
Very sorry to come back over here for answer, pardon me _/\\_
Just one quick question.
Does an epub file require internet connection for all of the (including images, links etc) contents to load?

**Comment by LingarajBagali** [comment]
> @LingarajBagali I'm not quite sure I understand your question. Do you want to load the content of each HTTP links of the eBook or just load images ?
> 
> Regarding HTML link I guess it's self explanatory, you require internet connection to get the content.
> 
> Regarding images :
> **As I am not familiar with `epub` format, all my following answers are assumptions.**
> 
> `epub` format is basically HTML, hence the possibility of putting images as HTTP link using `<img src="http://domain.org/path/to/image" />`. Still, it would require the eBook reader to implement an HTTP client, which I highly doubt about since mostly eBooks are meant to be read while offline. This lead me to think that it should be a bad practice to add an HTTP link to the image resource. A second reason to this is that links are not n everlasting resources and may perish through time, leading the eBook without its resource.
> 
> So I guess it can be possible to add online images to an `epub` file, but I highly doubt serious books does so.
> 
> But as I said, these are mostly assumptions and you should double check these through experience or existing documentation.
> Using calibre, I found [this documentation](https://manual.calibre-ebook.com/edit.html) which should let you try to add an HTTP link to an existing eBook to see if it would work.
> I don't know much about Windows similar tools but I doubt it's complicated to find.

Much appreciated mate @hypr2771 for all the information that you shared you knew of.
One thing, what about this script? does it create EPUB files that can be carried on any offline device and can be read? (i meant, when the script creates an EPUB file of one bookid, will it have all the contents of that book, including links, images etc etc without any miss?)

**Comment by LingarajBagali** [comment]
And, just as a feature request, it would be good if the script separately keeps all of the files (html, images, css etc) in a folder.

**Comment by nicohaenggi** [comment]
@LingarajBagali The way I developed it, it downloads all HTML and CSS files, and all images and includes them into the EPUB. So you won't need any internet access when trying to read the generated EPUB.

@hypr2771 That's right. The EPUB specification does allow to include external images but is usually avoided due to obvious reasons.

**Comment by hypr2771** [comment]
Just as I was curious, I tried using `calibre`'s `Edit book` tool.
![image](https://user-images.githubusercontent.com/6195718/51443807-c1c1ab80-1cee-11e9-9e9e-b209545e40aa.png)
It actually implements an HTTP client to fetch the images having an `<img src="http://domain.org/path/to/image" ...>`, so one can not guarantee every eBook resources will work offline, but you can consider those as poor quality eBook...
So I would not worry much about it @LingarajBagali.

**Comment by LingarajBagali** [comment]
> Just as I was curious, I tried using `calibre`'s `Edit book` tool.
> ![image](https://user-images.githubusercontent.com/6195718/51443807-c1c1ab80-1cee-11e9-9e9e-b209545e40aa.png)
> It actually implements an HTTP client to fetch the images having an `<img src="http://domain.org/path/to/image" ...>`, so one can not guarantee every eBook resources will work offline, but you can consider those as poor quality eBook...
> So I would not worry much about it @LingarajBagali.

ohh, is this an EPUB downloaded from this script that you tried to edit?

**Comment by hypr2771** [comment]
@LingarajVB to make everything crystal clear :
The eBook I downloaded worked perfectly fine with an embedded image in the eBook folder.
I wanted to check whether one can put an HTTP link instead of an image in the `epub` format.
To do so, I replaced a link to the static image in the static image folder of the eBook to an HTTP link.
`calibre` eBook client was able to fetch the image over the internet and show it in the eBook.

It was just an experiment to validate my previous assumption.

The eBook was indeed fully viewable offline before my modification (including images).
Once I modified it, only the modified image was not viewable due to the lack of internet connection to fetch the image over the internet.


**TL;DR : you can read any *good quality* eBook offline.**

**Comment by LingarajVB** [comment]
> @LingarajVB to make everything crystal clear :
> The eBook I downloaded worked perfectly fine with an embedded image in the eBook folder.
> I wanted to check whether one can put an HTTP link instead of an image in the `epub` format.
> To do so, I replaced a link to the static image in the static image folder of the eBook to an HTTP link.
> `calibre` eBook client was able to fetch the image over the internet and show it in the eBook.
> 
> It was just an experiment to validate my previous assumption.
> 
> The eBook was indeed fully viewable offline before my modification (including images).
> Once I modified it, only the modified image was not viewable due to the lack of internet connection to fetch the image over the internet.
> 
> **TL;DR : you can read any _good quality_ eBook offline.**

Great, Thanks for making me crystal clear now :) 👍
